lib/modules/manager/nix/extract.ts

```diff
@@ -1,23 +1,139 @@
+      continue;
+    }
+
+    // skip all locked nodes which are not in the flake.nix and cannot be updated
+    const rootInputs = flakeLock.nodes['root'].inputs;
+    if (!rootInputs || !(depName in rootInputs)) {
+      logger.debug(
+        { packageLockFile, error: flakeLockParsed.error },
+        `flake.lock file cannot find "root" node or reference to ${depName}`,
+      );
+      continue;
+    }
+
+    const flakeInput = flakeLock.nodes[depName];
+    const flakeLocked = flakeInput.locked;
+    const flakeOriginal = flakeInput.original;
+
+    // istanbul ignore if: if we are not in a root node then original and locked always exist which cannot be easily expressed in the type
+    if (flakeLocked === undefined || flakeOriginal === undefined) {
+      logger.debug(
+        { packageLockFile },
+        `Found empty flake input '${JSON.stringify(flakeInput)}', skipping`,
+      );
+      continue;
+    }
+
+    // indirect inputs cannot be reliable updated because they depend on the flake registry
+    if (flakeOriginal.type === 'indirect') {
+      continue;
+    }
+
+    switch (flakeLocked.type) {
+      case 'github':
+        deps.push({
+          depName,
+          currentValue: flakeOriginal.ref,
+          currentDigest: flakeLocked.rev,
+          replaceString: flakeLocked.rev,
+          datasource: GitRefsDatasource.id,
+          packageName: `https://${flakeOriginal.host ?? 'github.com'}/${flakeOriginal.owner}/${flakeOriginal.repo}`,
+        });
+        break;
+      case 'gitlab':
+        deps.push({
+          depName,
+          currentValue: flakeOriginal.ref,
+          currentDigest: flakeLocked.rev,
+          replaceString: flakeLocked.rev,
+          datasource: GitRefsDatasource.id,
+          packageName: `https://${flakeOriginal.host ?? 'gitlab.com'}/${flakeOriginal.owner}/${flakeOriginal.repo}`,
+        });
+        break;
+      case 'git':
+        deps.push({
+          depName,
+          currentValue: flakeOriginal.ref,
+          currentDigest: flakeLocked.rev,
+          replaceString: flakeLocked.rev,
+          datasource: GitRefsDatasource.id,
+          packageName: flakeOriginal.url,
+        });
+        break;
+      case 'sourcehut':
+        deps.push({
+          depName,
+          currentValue: flakeOriginal.ref,
+          currentDigest: flakeLocked.rev,
+          replaceString: flakeLocked.rev,
+          datasource: GitRefsDatasource.id,
+          packageName: `https://${flakeOriginal.host ?? 'git.sr.ht'}/${flakeOriginal.owner}/${flakeOriginal.repo}`,
+        });
+        break;
+      case 'tarball':
+        deps.push({
+          depName,
+          currentValue: flakeLocked.ref,
+          currentDigest: flakeLocked.rev,
+          replaceString: flakeLocked.rev,
+          datasource: GitRefsDatasource.id,
+          // type tarball always contains this link
+          packageName: (
+            flakeOriginal.url ?? /* istanbul ignore next */ ''
```

Comment:
why the fallback? isn't the url always defined?
```suggestion
            flakeOriginal.url!
```

Reply:
Yeah, unless people create artificial broken lock files.